query-engine/core/src/query_document/parser.rs

```diff
@@ -109,157 +105,237 @@ impl QueryDocumentParser {
+                        error_kind: QueryParserErrorKind::RequiredValueNotSetError,
+                    })),
+                }
             })
             .collect::<Vec<QueryParserResult<ParsedArgument>>>()
             .into_iter()
             .collect()
     }
 
-    /// Parses and validates a QueryValue against an InputType, recursively.
-    #[rustfmt::skip]
-    pub fn parse_input_value(value: QueryValue, input_type: &InputType) -> QueryParserResult<ParsedInputValue> {
-        // todo figure out what is up with enums
-        match (&value, input_type) {
-            // Handle optional inputs
-            (_, InputType::Opt(ref inner))                  => Self::parse_input_value(value, inner),
-
-            // Handle null inputs
-            (QueryValue::Null, InputType::Null(inner))          => Ok(ParsedInputValue::Single(PrismaValue::null(&**inner))),
-            (_, InputType::Null(ref inner))                 => Self::parse_input_value(value, inner),
-
-            // The optional handling above guarantees that if we hit a Null here, a required value is missing.
-            (QueryValue::Null, _)                           => Err(QueryParserError::RequiredValueNotSetError),
-
-            // Scalar and enum handling.
-            (_, InputType::Scalar(scalar))                  => Self::parse_scalar(value, &scalar).map(ParsedInputValue::Single),
-            (QueryValue::Enum(_), InputType::Enum(et))      => Self::parse_enum(value, et),
-            (QueryValue::String(_), InputType::Enum(et))    => Self::parse_enum(value, et),
-            (QueryValue::Boolean(_), InputType::Enum(et))   => Self::parse_enum(value, et),
-
-            // List and object handling.
-            (QueryValue::List(values), InputType::List(l))  => Self::parse_list(values.clone(), &l).map(ParsedInputValue::List),
-            (_, InputType::List(l))                         => Self::parse_list(vec![value], &l).map(ParsedInputValue::List),
-            (QueryValue::Object(o), InputType::Object(obj)) => Self::parse_input_object(o.clone(), obj.into_arc()).map(ParsedInputValue::Map),
-            (_, input_type)                                 => Err(QueryParserError::ValueTypeMismatchError { have: value, want: input_type.clone() }),
+    /// Parses and validates a QueryValue against possible input types.
+    /// Matching is done in order of definition on the input type. First matching type wins.
+    pub fn parse_input_value(
+        parent_path: QueryPath,
+        value: QueryValue,
+        possible_input_types: &[InputType],
+    ) -> QueryParserResult<ParsedInputValue> {
+        let mut parse_results = vec![];
+
+        for input_type in possible_input_types {
+            let value = value.clone();
+            let result = match (&value, input_type) {
+                // Null handling
+                (QueryValue::Null, InputType::Scalar(ScalarType::Null)) => {
+                    Ok(ParsedInputValue::Single(PrismaValue::null(TypeHint::Unknown)))
```

Comment:
This will be now translated to be a null integer from QE to Quaint, and will panic out if the underlying field is something that is not compatible with int in MSSQL. Are we actually converting this to something that follows the schema type in lower levels of QE?

Reply:
MSSQL is irrelevant for this PR. That said, the entire type hint handling will be removed and replaced with something that is on the SQL connector level, because it the abstractions are leaky and they're failing (as you correctly noted here).